File: CallbackTechniques/Task/Task.m
//
//  Task.m
//  PerformanceTests
//
//  Created by Alexey Naumov on 04/03/2017.
//  Copyright © 2017 Alexey Naumov. All rights reserved.
//

#import "Task.h"

@implementation ObjCOperationCaller

- (NSOperation *)operation
{
  return [NSBlockOperation blockOperationWithBlock:^{ }];
}

@end

@implementation ObjCOperationCallee

- (NSOperation *)operationAwaiting:(NSOperation *)operationToWait
{
  __weak typeof(self) weakSelf = self;
  NSOperation *operation = [NSBlockOperation blockOperationWithBlock:^{
    weakSelf.wasCalled = YES;
  }];
  [operation addDependency:operationToWait];
  return operation;
}

@end

File: CallbackTechniques/Task/Task.swift
//
//  Task.swift
//  PerformanceTests
//
//  Created by Alexey Naumov on 04/03/2017.
//  Copyright © 2017 Alexey Naumov. All rights reserved.
//

class SwiftOperationCaller {
  
  func operation() -> Operation {
    return BlockOperation(block: { })
  }
  
}

class SwiftOperationCallee {
  
  var wasCalled = false
  
  func operation(awaiting operationToWait: Operation) -> Operation {
    let operation = BlockOperation(block: { [weak self] _ in
      self?.wasCalled = true
    })
    operation.addDependency(operationToWait)
    return operation
  }
  
}
